File: Solution_coinchange.py
"""
O(S*N) where S is amount between 1<=i<=S where N is no of coin denomination
O(S+N) extra space for dp table

-initialise dp table as list of list and assign inf or higher value to Ast row and for 0 to amount columns
-if the curr amt less than denomination , copy the previous row same col value 
-else find min between (prev value,1+dp[i][j-curr_denomination]
-if value gt than 9999 or Inf return -1
"""


class Solution_coinchange:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if coins is None or len(coins)==0:
            return 0
        dp = [[0 for i in range(amount+1)] for j in range(len(coins)+1)] #integer overflow is handled in pytho so we can put infinity or 9999 or amount+1
        dp[0]= [9999]*(amount+1)
        dp[0][0] = 0
        for i in range(1,len(coins)+1):
            for j in range(1,amount+1):
                if(j<coins[i-1]):
                    dp[i][j]=dp[i-1][j]
                else:
                    dp[i][j]=min(dp[i-1][j],1+dp[i][j-coins[i-1]])
        if(dp[len(coins)][amount]>=9999):
            return -1
        else:
            return dp[len(coins)][amount]#last row last column
                
                    
                
        

File: Solution_coinchange.java
/*
//Time Complexity :  O(n * m) , here n and m is the number of row and number of column in dp matrix.

//Space Complexity :  O(n * m) here n and m is the number of row and number of column in dp matrix.

// Did this code successfully run on Leetcode : Yes, worked on the leetcode

// Any problem you faced while coding this : I had faced challenge in created a condition 
for the choose case and at last I forgot to check for the if we didn't get minimum coin then return -1.

*/

public class Solution_coinchange {

    public int coinChange(int[] coins, int amount) {

        return helper(coins, amount, 0, 0);

    }

    private int helper(int[] coins, int amount, int index, int count) {

        // creating dp array with adding one more size for if our coins has only 1
        // value.

        int[][] dp = new int[coins.length + 1][amount + 1];

        // we can't get any amount with any coin so adding maximum amount.

        for (int j = 1; j < dp[0].length; j++) {

            dp[0][j] = amount + 1;

        }

        // Itreating throught the dp array.

        for (int i = 1; i < dp.length; i++) {

            for (int j = 1; j < dp[0].length; j++) {

                // checking if amount is >= then coin
                if (j >= coins[i - 1]) {

                    dp[i][j] = Math.min(dp[i - 1][j], 1 + dp[i][j - coins[i - 1]]);
                }
                // if amount is < coin then just use value above that row and same column.
                else {

                    dp[i][j] = dp[i - 1][j];
                }
            }
        }

        // checking if last element value has grether than amount then returning -1 else
        // value.
        return dp[dp.length - 1][dp[0].length - 1] == amount + 1 ? -1 : dp[dp.length - 1][dp[0].length - 1];

    }

}
